### PR TITLE
Warn about forgetting attention mask functions

### PR DESCRIPTION
**Pull request by TuringTux**
As it is implemented currently, when registering a custom attention function, but not a custom attention mask function, the mask is set to `None`.

This can cause unexpected issues, as it means that e.g., causal masking is silently discarded when you only try to change a minor detail of the attention implementation of an existing model, with grave changes to the overall performance (the model starts to cheat). This happened to me in one of my recent projects.

## What does this PR do?

To hopefully save future readers of the docs that problem, this PR adds:

1. An explicit warning that a mask function should likely be registered
2. Copyable code to register the sdpa mask as a default (so that if you just copy/paste code from the docs, you shouldn't fall victim to this issue).

## Code Agent Policy

- [x] I confirm that this is not a pure code agent PR.

(in fact, no LLM whatsoever were involved in creating this PR)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?

## Who can review?

@stevhliu, maybe?

### CODE REVIEW

File: docs/source/en/attention_interface.md [hunk 1]
```diff
@@ -130,18 +130,23 @@ model = AutoModelForImageTextToText.from_pretrained(
 
 Customize or create new attention functions by adding them to the attention registry with [`AttentionInterface.register`]. Models use these functions through the `attn_implementation` argument.
 
-This example customizes the attention function to print a statement for each layer.
+> [!WARNING]
+> If you register a new attention function, you likely also need to register a new attention _mask_ function. If you do not register a new attention _mask_ function, **the mask is set to `None`, i.e., completely ignored**. This could break your model in unexpected ways (e.g., by silently discarding Causal Masking).
```
Comment (stevhliu) [inline docs/source/en/attention_interface.md]: i think this wording may be a bit more accurate?

```suggestion
> [!WARNING]
> Register a matching attention mask function when you register a custom attention function. If the custom `attn_implementation` name is not registered in [`AttentionMaskInterface`], Transformers skips mask creation and passes `attention_mask=None` to the attention layers. Your attention function must handle causal, padding, packing, or sliding-window constraints itself, or those constraints can be silently dropped.
```

File: docs/source/en/attention_interface.md [hunk 2]
```diff
@@ -130,18 +130,23 @@ model = AutoModelForImageTextToText.from_pretrained(
...
+
+This example customizes the attention function to print a statement for each layer (and keeps whatever mask is used in the original implementation, by registering [`masking_utils.sdpa_mask`] as attention mask function):
```
Comment (stevhliu) [inline docs/source/en/attention_interface.md]:
```suggestion
This example customizes the attention function to print a statement for each layer. It keeps the mask in the original implementation by registering `masking_utils.sdpa_mask` as the attention mask function.
```

**Review by stevhliu — APPROVED** [review]
thanks! cc @Cyrilvallez for a look

File: docs/source/en/attention_interface.md (hunk 1 above)
Comment (TuringTux) [reply docs/source/en/attention_interface.md]: Definitely, thanks!

I have reworded the suggestion a bit, to make it more clear to me when which consequences are to be expected; however, feel free to change it back to your original suggestion or modify it further as you see fit.

**Comment by TuringTux** [comment]
Thanks for the fast review!

**Comment by Cyrilvallez** [comment]
Makes sense, thanks @TuringTux!
@stevhliu I updated according to your last suggestion, as this was a bit overlooked before. Please have a quick look to make sure wording is to your liking!
